design-proposals/klip-7-connect-integration.md

```diff
@@ -0,0 +1,205 @@
+
+- A single connector can have many topics, which makes modeling them as a single statement difficult
+- A connector can be modeled in different ways (e.g. stream/table) and connector lifecycle should
+  be independent with how KSQL interprets it
+
+In the future, there can exist shortcuts (similar to `C*AS`) to both create connectors and declare 
+corresponding streams/tables.
+
+There are a few more minor points of consideration:
+- All "DDL" commands that cause changes in connect will not be submitted to the command topic;
+  instead they will be issued on the REST server (this follows the pattern that all external-state
+  modifications are not distributed)
+- To make this work for all vanilla connectors, we will need Connect to implement an API that shows
+  all topics created by a connector.
+
+**Sample Ingress**
+
+The first step is to create and start a Connector via the REST API (`POST /connectors`)
+```sql
+ksql> CREATE EXTERNAL SOURCE my-postgres-jdbc WITH ( \
+      connector.class="io.confluent.connect.jdbc.jdbcSourceConnector", \ 
+      tasks.max=1, \
+      connection.url=jdbc:postgresql://localhost:5432/my-db, \
+      ...);
+
+--------------
+SOURCE CREATED
+--------------
+```
+*Note: as Connect manages to simplify the configuration, whether through templates or any other
+means, we will immediately be able to leverage this simplification in KSQL's WITH clause.*
+
+Next, we provide tools to understand the status of the connector and corresponding topics hitting
+the `GET /connectors` and `GET /connectors/<name>` REST endpoints:
+```sql
+ksql> SHOW EXTERNAL SOURCES;
+
+    name         |              connector                        |   topic prefix
+-----------------|-----------------------------------------------|--------------------
+my-postgres-jdbc | io.confluent.connect.jdbc.JdbcSourceConnector | test-postgres-jdbc
+
+ksql> DESCRIBE my-postgres-jdbc;
+
+name: my-postgres-jdbc
+connector: io.confluent.connect.jdbc.SourceConnector
+topic-prefix: test-postgres-jdbc
+
+   worder id     | task | status
+-----------------|------|---------
+10.200.7.69:8083 | 0    | RUNNING
+
+ name      |     topic                     |  schema
+-----------|-------------------------------|--------------------------------------
+user-table | test-postgres-jdbc-user-table | id VARCHAR, user VARCHAR, age INTEGER
+page-table | test-postgres-jdbc-page-table | id VARCHAR, page VARCHAR, url VARCHAR
+```
+
+Then the user can create streams or tables from the underlying topics:
+```sql
+ksql> CREATE TABLE users_original WITH (kafka_topic = 'test-postgres-jdbc-user-table', ...);
+
+--------------
+STREAM CREATED
+--------------
+```
+
+At this point, assuming the `pageviews_original` stream is already created, we can join the newly
+created `users_original` table with the `pageviews_original` stream. Finally, the user can drop and 
+terminate the connectors.
+```sql
+ksql> DROP EXTERNAL SOURCE my-postgres-jdbc TERMINATE;
```

Comment:
What does the `TERMINATE` at the end of this statement signify? Will there be an option to drop an external source without terminating the connector?

Reply:
I was imagining you'd be able to remove the metadata for a connector without actually terminating the connector - though I suppose at second though I'm not sure why that would be especially useful.